draft-ietf-quic-tls.md

```diff
@@ -255,142 +205,169 @@ A simplified TLS 1.3 handshake with 0-RTT application data is shown in
-* "{" and "}" enclose messages that are protected by the TLS Handshake keys.
-
-If 0-RTT is not attempted, then the client does not send packets protected by
-the 0-RTT key (@0).  In that case, the only key transition on the client is from
-handshake packets (@H) to 1-RTT protection (@1), which happens after it sends
-its final set of TLS handshake messages.
-
-Note: two different types of packet are used during the handshake by both client
-and server.  The Initial packet carries a TLS ClientHello message; the remainder
-of the TLS handshake is carried in Handshake packets.  The Retry packet carries
-a TLS HelloRetryRequest, if it is needed, and Handshake packets carry the
-remainder of the server handshake.
-
-The server sends TLS handshake messages without protection (@H).  The server
-transitions from no protection (@H) to full 1-RTT protection (@1) after it sends
-the last of its handshake messages.
-
-Some TLS handshake messages are protected by the TLS handshake record
-protection.  These keys are not exported from the TLS connection for use in
-QUIC.  QUIC packets from the server are sent in the clear until the final
-transition to 1-RTT keys.
-
-The client transitions from handshake (@H) to 0-RTT keys (@0) when sending 0-RTT
-data, and subsequently to to 1-RTT keys (@1) after its second flight of TLS
-handshake messages.  This creates the potential for unprotected packets to be
-received by a server in close proximity to packets that are protected with 1-RTT
-keys.
-
-More information on key transitions is included in {{hs-protection}}.
+{: #schematic title="QUIC and TLS Interactions"}
 
+Unlike TLS over TCP, QUIC applications which want to send data do not
+send it through TLS "application_data" records. Rather, they send it
+as QUIC STREAM frames which are then carried in QUIC packets.
+
+
+# Carrying TLS Messages {#carrying-tls}
+
+QUIC carries TLS handshake data in CRYPTO_HS frames, each of which
+consists of a contiguous block of handshake data identified by an
+offset and length. Those frames are packaged into QUIC packets
+and encrypted under the current TLS encryption level.
+As with TLS over TCP, once TLS handshake data has
+been delivered to QUIC, it is QUIC's responsibility to deliver it
+reliably. Each chunk of data is associated with the then-current TLS
+sending keys, and if QUIC needs to retransmit that data, it MUST use
+the same keys even if TLS has already updated to newer keys.
+
+One important difference between TLS 1.3 records (used with TCP)
+and QUIC CRYPTO_HS frames is that in QUIC multiple frames may appear
+in the same QUIC packet as long as they are associated with the
+same encryption level. For instance, an implementation might
+bundle a Handshake message and an ACK for some Handshake
+data into the same packet.
+
+In general, the rules for which data can appear in packets of which
+encryption level are the same in QUIC as in TLS over TCP:
+
+- CRYPTO_HS frames MAY appear in packets of any encryption level.
```

Comment:
I'm not sure where to point this out, but shouldn't we have a restriction saying that a QUIC packet can only have *one* CRYPTO_HS frame?

Reply:
Why? That's not, for instance, how Minq works.

Reply:
Because it's like having two STREAM frames for the same stream ID.

Reply:
This also seems to be permitted for STREAM frames:

```
   Stream multiplexing is achieved by interleaving STREAM frames from
   multiple streams into one or more QUIC packets.  A single QUIC packet
   can include multiple STREAM frames from one or more streams.
```

Reply:
There's nothing wrong with having multiple STREAM frames from the same stream, and I don't see anything wrong with having multiple CRYPTO_HS frames for the same reason.  Presumably, they contain different regions of the handshake stream generated at different points by the TLS stack.

(If they claim to contain the same region of data but differ, that might be a good reason to blow up, of course....)

Reply:
Right, so I get the case where more than one CRYPTO_HS/STREAM frame is present but at least we should mention what to do if they overlap.

Reply:
I think we should, but not limit it to the case where they are in the same packet. The situation is the same if they are in two different packets.

Reply:
If they are in two different packets then it's just a retransmission ;-)
I'm thinking we just need some text saying the receive is supposed to detect and handle overlapping CRYPTO_HS/STREAM frame even if they are in the same packet.  Maybe we already have that, but I don't see it.

Reply:
Sure. Given that CRYPTO_HS has the same semantics in this respect as STREAM, I suggest a separate PR that adds it to the stream section.

Reply:
@huitema:
> I am busy implementing this, and I have a small issue with the "MAY appear in packets of any encryption level." How do I verify that the packets are sent as intended? For example, what if I receive a CRYPTO_HS frame with offset 500 in an handshake-protected, and another CRYPTO_HS frame with an overlapping offset in an application-data protected packet?

My assumption is that such check will be provided as an API or a guideline by the TLS stack, and that therefore it is unnecessary to clarify in the specification on how that should be done.

For example, the `ptls_handle_messages` function of [picotls](https://github.com/h2o/picotls/pull/141) accepts the epoch as the argument, and returns an error if a TLS messages was being received in an incorrect epoch.

Reply:
@huitema: I don't understand what you are asking here, because those are different data namespaces.

So, for instance, SH will be sent with Handshake[Offset=0], and NST will be sent with Application[Offset=0]. They don't interact with each other.

The case that is an error is when data comes in at a given encryption level after the TLS state machine has moved on. For instance, if the server sends Handshake: [Offset=X, SFIN], Handshake: [Offset=X + len(SFIN), <something>]. This isn't valid because the TLS handshake is now reading at Application. As Kazuho says, this should generate an error, and the text requires that:

```
- If the packet is from a previously installed encryption level, it
  MUST not contain data which extends past the end of previously
  received data in that flow. Implementations MUST treat any
  violations of this requirement as a connection error of type
  PROTOCOL_VIOLATION.
```